src/web_clipboard.rs

```diff
@@ -0,0 +1,106 @@
+use std::sync::{
+    mpsc::{self, Receiver, Sender},
+    Arc, Mutex,
+};
+
+use bevy::prelude::*;
+use wasm_bindgen_futures::spawn_local;
+
+use crate::EguiClipboard;
+
+/// startup system for bevy to initialize clipboard.
+pub fn startup(mut clipboard_channel: ResMut<EguiClipboard>) {
+    setup_clipboard_paste(&mut clipboard_channel.clipboard)
+}
+
+fn setup_clipboard_paste(clipboard_channel: &mut WebClipboardPaste) {
+    let (tx, rx): (Sender<String>, Receiver<String>) = mpsc::channel();
+
+    use wasm_bindgen::closure::Closure;
+    use wasm_bindgen::prelude::*;
+
+    let closure = Closure::<dyn FnMut(_)>::new(move |event: web_sys::ClipboardEvent| {
+        // TODO: maybe we should check if current canvas is selected ? not sure it's possible,
+        // but reacting to event at the document level will lead to problems if multiple games are on the same page.
+        match event
+            .clipboard_data()
+            .expect("could not get clipboard data.")
+            .get_data("text/plain")
+        {
+            Ok(data) => {
+                tx.send(data);
+            }
+            _ => {
+                info!("Not implemented.");
+            }
+        }
+        info!("{:?}", event.clipboard_data())
+    });
+
+    // TODO: a lot of unwraps ; it's using documents because paste event set on a canvas do not trigger (also tested on firefox in vanilla javascript)
+    web_sys::window()
+        .unwrap()
+        .document()
+        .unwrap()
+        .add_event_listener_with_callback("paste", closure.as_ref().unchecked_ref())
+        .expect("Could not edd paste event listener.");
+    closure.forget();
+    *clipboard_channel = WebClipboardPaste {
+        rx: Some(Arc::new(Mutex::new(rx))),
+    };
+
+    info!("setup_clipboard_paste OK");
+}
+
+/// To get data from web paste events
+#[derive(Default)]
+pub struct WebClipboardPaste {
+    rx: Option<Arc<Mutex<Receiver<String>>>>,
```

Comment:
I'd be happier if we used `NonSend` or crossbeam channels (supporting `Sync`) instead of a `Mutex`

Reply:
I went for crossbeam channel, NonSend is a lifetime error fest and has to change egui res to NonSendMut